pkg/virtualKubelet/reflection/storage/controller.go

```diff
@@ -0,0 +1,196 @@
+	if claim.Spec.VolumeName != "" {
+		return false, nil
+	}
+
+	if provisioner, found := claim.Annotations[annStorageProvisioner]; found {
+		if npvcr.knownProvisioner(provisioner) {
+			claimClass := util.GetPersistentVolumeClaimClass(claim)
+			if claimClass != npvcr.virtualStorageClassName {
+				return false, nil
+			}
+
+			class, err := npvcr.classes.Get(claimClass)
+			if err != nil {
+				return false, err
+			}
+
+			if class.VolumeBindingMode != nil && *class.VolumeBindingMode == storagev1.VolumeBindingWaitForFirstConsumer {
+				// When claim is in delay binding mode, annSelectedNode is
+				// required to provision volume.
+				// Though PV controller set annStorageProvisioner only when
+				// annSelectedNode is set, but provisioner may remove
+				// annSelectedNode to notify scheduler to reschedule again.
+				if selectedNode, ok := claim.Annotations[annSelectedNode]; ok && selectedNode != "" {
+					return true, nil
+				}
+				return false, nil
+			}
+			// do not provision if the binding mode is not WaitForFirstConsumer
+			return false, nil
+		}
+	}
+
+	return false, nil
+}
+
+func (npvcr *NamespacedPersistentVolumeClaimReflector) knownProvisioner(provisioner string) bool {
+	return provisioner == npvcr.provisionerName
+}
+
+type provisionFunc func(context.Context, controller.ProvisionOptions) (*v1.PersistentVolume, controller.ProvisioningState, error)
+
+// provisionClaimOperation attempts to provision a volume for the given claim.
+// Returns nil error only when the volume was provisioned (in which case it also returns ProvisioningFinished),
+// a normal error when the volume was not provisioned and provisioning should be retried (requeue the claim),
+// or the special errStopProvision when provisioning was impossible and no further attempts to provision should be tried.
+func (npvcr *NamespacedPersistentVolumeClaimReflector) provisionClaimOperation(ctx context.Context,
+	claim *v1.PersistentVolumeClaim, provision provisionFunc) (controller.ProvisioningState, error) {
+	// Most code here is identical to that found in controller.go of kube's PV controller...
+	claimClass := util.GetPersistentVolumeClaimClass(claim)
+
+	//  A previous doProvisionClaim may just have finished while we were waiting for
+	//  the locks. Check that PV (with deterministic name) hasn't been provisioned
+	//  yet.
+	pvName := strings.Join([]string{"pvc", string(claim.UID)}, "-")
+	_, err := npvcr.volumes.Get(pvName)
+	if err == nil {
+		// Volume has been already provisioned, nothing to do.
+		klog.V(4).Infof("Persistentvolume %q already exists, skipping", pvName)
+		return controller.ProvisioningFinished, nil
+	}
+
+	// Prepare a claimRef to the claim early (to fail before a volume is
+	// provisioned)
+	claimRef, err := ref.GetReference(scheme.Scheme, claim)
+	if err != nil {
+		klog.Error("Unexpected error getting claim reference: %v", err)
```

Comment:
Could be worth adding the failing PVC name in the log message (same for the ones below).